pkg/cloud/services/networking/router.go

```diff
@@ -239,19 +235,76 @@ func (s *Service) getRouterInterfaces(routerID string) ([]ports.Port, error) {
 	return portList, nil
 }
 
-func (s *Service) existsRouter(routerID string) (bool, error) {
-	if routerID == "" {
-		return false, nil
+type Router struct {
+	router routers.Router
+	subnet subnets.Subnet
+}
+
+func (s *Service) getRouter(clusterName string) (*Router, error) {
+	routerName := getRouterName(clusterName)
+	router, err := s.getRouterByName(routerName)
+	if err != nil {
+		return nil, err
+	}
+
+	subnetName := getSubnetName(clusterName)
+	subnet, err := s.getSubnetByName(subnetName)
+	if err != nil {
+		return nil, err
 	}
-	router, err := routers.Get(s.client, routerID).Extract()
+	return &Router{router: router, subnet: subnet}, nil
+}
+
+func (s *Service) getRouterByName(routerName string) (routers.Router, error) {
+	allPages, err := routers.List(s.client, routers.ListOpts{
+		Name: routerName,
+	}).AllPages()
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return false, nil
-		}
-		return false, err
+		return routers.Router{}, err
+	}
+
+	routerList, err := routers.ExtractRouters(allPages)
+	if err != nil {
+		return routers.Router{}, err
+	}
+
+	if len(routerList) > 1 {
+		return routers.Router{}, fmt.Errorf("found %d router with the name %s, which should not happen", len(routerList), routerName)
+	}
+
+	switch len(routerList) {
+	case 0:
+		return routers.Router{}, nil
+	case 1:
+		return routerList[0], nil
+	}
+	return routers.Router{}, errors.New("too many resources")
```

Comment:
Can you move line 272 here? Because that's the same case and so you don't have to handle this case twice and check the list length
```suggestion
	return routers.Router{}, fmt.Errorf("found %d router with the name %s, which should not happen", len(routerList), routerName)
```